File: dash_docs/chapters/advanced_callbacks/examples/last_clicked_button.py
import json

import dash
import dash_html_components as html
from dash.dependencies import Input, Output

app = dash.Dash(__name__)

app.layout = html.Div([
    html.Button('Button 1', id='btn-1'),
    html.Button('Button 2', id='btn-2'),
    html.Button('Button 3', id='btn-3'),
    html.Div(id='container')
])


@app.callback(Output('container', 'children'),
              [Input('btn-1', 'n_clicks'),
               Input('btn-2', 'n_clicks'),
               Input('btn-3', 'n_clicks')])
def display(btn1, btn2, btn3):
    ctx = dash.callback_context

    if not ctx.triggered:
        button_id = 'No clicks yet'
    else:
        button_id = ctx.triggered[0]['prop_id'].split('.')[0]

    ctx_msg = json.dumps({
        'states': ctx.states,
        'triggered': ctx.triggered,
        'inputs': ctx.inputs
    }, indent=2)

    return html.Div([
        html.Table([
            html.Tr([html.Th('Button 1'),
                     html.Th('Button 2'),
                     html.Th('Button 3'),
                     html.Th('Most Recent Click')]),
            html.Tr([html.Td(btn1 or 0),
                     html.Td(btn2 or 0),
                     html.Td(btn3 or 0),
                     html.Td(button_id)])
        ]),
        html.Pre(ctx_msg)
    ])


if __name__ == '__main__':
    app.run_server(debug=True)

File: dash_docs/chapters/faq_gotchas/examples/last_clicked_button.R
library(dash)
library(dashCoreComponents)
library(dashHtmlComponents)
library(jsonlite)

app <- Dash$new()

app$layout(
  htmlDiv(
    list(
      htmlButton(children = 'Button 1', id='btn-1', n_clicks=0),
      htmlButton(children = 'Button 2', id='btn-2', n_clicks=0),
      htmlButton(children = 'Button 3', id='btn-3', n_clicks=0),
      htmlDiv(id='container')
      
    )
  )
)

app$callback(output('container', 'children'),
             list(input('btn-1', 'n_clicks'),
                  input('btn-2', 'n_clicks'),
                  input('btn-3', 'n_clicks')),
    function(input1, input2, input3) {
      
      inputs <- c(input1, input2, input3)
      
      ctx <- app$callback_context()
      
      most_recent_click <- if(ctx$triggered$value) {
        
        unlist(strsplit(ctx$triggered$prop_id, "[.]"))[1]
        
      } else "No clicks yet"

      toJson <- jsonlite::toJSON(ctx, pretty = TRUE)
      
      htmlDiv(
        list(
          htmlTable(
            list(
              htmlTr(
                list(
                  htmlTh('Button 1'),
                  htmlTh('Button 2'),
                  htmlTh('Button 3'),
                  htmlTh('Most Recent Click')
                )
              ),
              htmlTr(
                list(
                  htmlTd(input1),
                  htmlTd(input2),
                  htmlTd(input3),
                  htmlTd(most_recent_click)
                )
              )
            )
          ),
          htmlPre(toJson)
        )
      )
    })
#app$run_server()
